core/src/main/java/org/elasticsearch/search/SearchService.java

```diff
@@ -792,6 +793,17 @@ private void parseSource(DefaultSearchContext context, SearchSourceBuilder sourc
             }
             context.storedFieldsContext(source.storedFields());
         }
+
+        if (source.collapse() != null) {
```

Comment:
Cool, that is better than where is initially was.

I think the downside of modifying the search context compared to the search source builder is that the query gets logged on the ES side (slow query log) then it isn't obvious that this request also fetches doc values. Someone would need to know that this is what collapse does. So to me adding this logic here is less transparent compared than in `SearchSourceBuilder`. What do you think?